File: Leetcode/22_Generate_Parentheses.js
var res = [];
var stack = [];
// n=3

function backTrack(openN, closeN) {
    if (openN == n && closeN == n) {
        res.push(stack.join(""));
        return;
    }

    if (openN < n) {
        stack.push("(");
        backTrack(openN + 1, closeN);
        stack.pop()
    }

    if(closeN < openN) {
        stack.push(")");
        backTrack(openN, closeN + 1);
        stack.pop()
    }
}

backTrack(0, 0)
console.log(res)

File: Leetcode/22_Generate_Parentheses.py
from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        stack = []
        res = []

        def backTrack(openN, closeN):
            print(stack, openN, closeN)
            if openN == closeN == n:
                res.append("".join(stack))
                return

            if openN < n:
                stack.append("(")
                backTrack(openN + 1, closeN)
                stack.pop()

            if closeN < openN:
                stack.append(")")
                backTrack(openN, closeN + 1)
                stack.pop()

        backTrack(0, 0)
        return res


s = Solution()
print(s.generateParenthesis(3))  # output ["((()))","(()())","(())()","()(())","()()()"]
# print(s.generateParenthesis(1))  # output ["()"]


